File: pkg/reporting/reporter.go
package reporting

import (
	"context"
	"encoding/json"
	"log/slog"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	"github.com/cmsgov/emmy-api/pkg/core"
)

type ReportData struct {
	Endpoint   string    `json:"endpoint"`
	DataSource string    `json:"data_source"`
	ClientID   string    `json:"client_id"`
	Success    bool      `json:"success"`
	Timestamp  time.Time `json:"timestamp"`
	StatusCode int       `json:"status_code"`
}

type Reporter interface {
	Report(ctx context.Context, data *ReportData)
}

type reporter struct {
	logger    *slog.Logger
	sqsClient *sqs.Client
	cfg       core.ReportingConfig
}

func NewReporter(ctx context.Context, cfg core.ReportingConfig, logger *slog.Logger) Reporter {
	if logger == nil {
		logger = slog.Default()
	}
	logger = logger.With(slog.String("component", "reporting"))

	var sqsClient *sqs.Client
	if cfg.SQSQueueURL != "" {
		awsCfg, err := config.LoadDefaultConfig(ctx)
		if err != nil {
			logger.Error("failed to load AWS config", "error", err)
		} else {
			sqsClient = sqs.NewFromConfig(awsCfg)
		}
	}

	return &reporter{
		cfg:       cfg,
		logger:    logger,
		sqsClient: sqsClient,
	}
}

func (r *reporter) Report(ctx context.Context, data *ReportData) {
	if r.cfg.SQSQueueURL != "" && r.sqsClient != nil {
		body, err := json.Marshal(data)
		if err != nil {
			r.logger.ErrorContext(ctx, "failed to marshal report data", "error", err)
			return
		}

		_, err = r.sqsClient.SendMessage(ctx, &sqs.SendMessageInput{
			QueueUrl:    aws.String(r.cfg.SQSQueueURL),
			MessageBody: aws.String(string(body)),
		})
		if err != nil {
			r.logger.ErrorContext(ctx, "failed to send report to SQS", "error", err)
			return
		}
		r.logger.DebugContext(ctx, "report sent to SQS", "queue_url", r.cfg.SQSQueueURL)
	} else {
		r.logger.InfoContext(ctx, "api call report",
			slog.String("endpoint", data.Endpoint),
			slog.Bool("success", data.Success),
			slog.String("data_source", data.DataSource),
			slog.String("client_id", data.ClientID),
			slog.Time("timestamp", data.Timestamp),
			slog.Int("status_code", data.StatusCode),
		)
	}
}

File: rails-app/app/services/reporting/reporter.rb
require "aws-sdk-sqs"

module Reporting
  class Reporter
    def initialize
      @queue_url = ENV["SQS_QUEUE_URL"]

      if @queue_url.present?
        @sqs_client = Aws::SQS::Client.new
      end
    end

    def report(data)
      if @queue_url.present? && @sqs_client
        begin
          @sqs_client.send_message(
            queue_url: @queue_url,
            message_body: data.to_json
          )
          Rails.logger.debug("report sent to SQS: #{@queue_url}")
        rescue StandardError => e
          Rails.logger.error("failed to send report to SQS: #{e.message}")
        end
      else
        Rails.logger.info(
          "api call report: endpoint=#{data.endpoint} success=#{data.success} " \
          "data_source=#{data.data_source} client_id=#{data.client_id} " \
          "timestamp=#{data.timestamp} status_code=#{data.status_code}"
        )
      end
    end
  end
end
